packages/run-protocol/src/vaultFactory/liquidateIncrementally.js

```diff
@@ -0,0 +1,322 @@
+/* eslint-disable no-await-in-loop */
+// no @ts-check
+
+import { E } from '@endo/eventual-send';
+import {
+  getAmountOut,
+  makeRatio,
+  assertProposalShape,
+  offerTo,
+  natSafeMath as NatMath,
+} from '@agoric/zoe/src/contractSupport/index.js';
+import { AmountMath } from '@agoric/ertp';
+import { Far } from '@endo/marshal';
+
+import { ceilMultiplyBy } from '@agoric/zoe/src/contractSupport/ratio';
+import { makeTracer } from '../makeTracer.js';
+// import { quote } from '@agoric/assert';
+
+const { details: X } = assert;
+const trace = makeTracer('LiqI');
+
+/**
+ * This contract liquidates a payment in multiple tranches to limit the
+ * price impact an the AMM of any one sale. Each block it will compute
+ * a tranche of collateral to sell, where the size is a function of
+ * the amount of that collateral in the AMM pool and the desired price impact.
+ * It then gets 3 prices for the current tranche:
+ * - AMM quote - compute xYK locally based on the pool sizes
+ * - Reserve quote - based on a low price at which the Reserve will purchase
+ * - Oracle quote - provide by the oracle that triggered liquidation
+ * - Oracle limit - a limit that is DIVERGENCE_TOLERANCE below the
+ *   Oracle quote.
+ * These sell based on the following:
+ * - If the AMM quote is *below* the oracle limit, skip this block
+ *   to wait for oracle and AMM prices to converge.
+ *    - should it still sell to teh Reserve here?
+ * - If the Reserve quote is higher than the AMM quote, sell to the
+ *   Reserve with a `want` of the Oracle limit
+ * - Otherwise, sell to the AMM with a `want` of the Oracle limit
+ * If selling a tranche for the minimum would pay off the debt (`isLast`),
+ * then use sell using`swapOut` with a minimum of the debt. If that fails
+ * this is not the last tranche, tehn sell with the `swapIn` operation,
+ * using the oracle price reference as the minimum.
+ *
+ * TODO integrate the reserve
+ */
+
+/** @type {ContractStartFn} */
+const start = async zcf => {
+  const { amm, priceAuthority, timerService, debtBrand } =
+    /** @type {LiquidationContractTerms} */ zcf.getTerms();
+  const nextBlock = async () => E(timerService).delay(1n);
+
+  // TODO make subject to governance
+  const SCALE = 1_000_000n;
+  const BASIS_POINTS = 10_000n * SCALE;
+  const BP2 = BASIS_POINTS * BASIS_POINTS;
```

Comment:
BP2 is 10**20. that's surprising.

I think if you want to use a denominator with higher resolution than Basis Points, it doesn't make sense to use BP_squared. SCALE (=10^6) would be fine, but this code is confusing me when numerators and denominators both have _BP appended to them.

When computing ratios, 30 BP is  30 / 10_000.   If you want something at higher resolution, say a half basis point, it would make sense to think of that as 50 / SCALE.  But multiplying the numerator by SCALE and the denominator by SCALE^2 is a surprising representation.

And mixing SCALE and BP2 is hard to follow.